drafts/str4d-overwinter-activation/draft1.rst

```diff
@@ -0,0 +1,196 @@
+      // Pre-Overwinter consensus rules applied to block
+  }
+
+
+Block parsing
+`````````````
+Incoming blocks known to have a particular height (due to their parent chain being entirely known) MUST be
+parsed under the consensus rules corresponding to their height.
+
+Incoming blocks with unknown heights (because at least one block in their parent chain is unknown) MUST NOT be
+considered valid, but MAY be cached for future consideration after all their parents have been received.
+
+Chain reorganization
+````````````````````
+It is possible for a reorganization to occur that rolls back from after the activation height, to before that
+height. This can handled in the same way as any regular chain orphaning or reorganization, as long as the new
+chain is valid over the same epochs.
+
+Post-activation upgrading
+`````````````````````````
+If a user does not upgrade their node to a compatible software version before ``ACTIVATION_HEIGHT`` is
+reached, their node will follow the pre-upgrade chain and download blocks that are incompatible with the
+post-upgrade branch. If the user subsequently upgrades their node to a compatible software version, the node
+will consider these blocks to be invalid, and MUST take one of the two following actions:
+
+- Discard all blocks of height ``ACTIVATION_HEIGHT`` and above, and then synchronize with the network.
+
+- Shut down and alert the user of the issue. In this case, the node could offer an option to perform the first
+  action.
+
+Memory pool
+-----------
+
+While the current chain tip height is below ``ACTIVATION_HEIGHT``, nodes SHOULD NOT accept transactions that
+will only be valid on the post-upgrade chain.
+
+When the current chain tip height reaches ``ACTIVATION_HEIGHT``, the node's local transaction memory pool
+SHOULD be cleared.
+
+Two-way replay protection
+-------------------------
+
+Before the Overwinter network upgrade, two-way replay protection is ensured by enforcing post-upgrade that the
+MSB of the transaction version is set to 1. From the perspective of old nodes, the transactions will have a
+negative version number, which is invalid under the old consensus rules. Enforcing this rule trivially makes
+old transactions invalid on the Overwinter branch.
+
+After the Overwinter network upgrade, two-way replay protection is ensured by transaction signatures
+committing to a specific ``BRANCH_ID``. [#zip-0143]_
+
+Wipe-out protection
+-------------------
+
+Nodes running upgrade-aware software versions will enforce the upgraded consensus rules from
+``ACTIVATION_HEIGHT``. The chain from that height will not reorg to a pre-upgrade branch if any block would
+violate the new consensus rules (such as including any old-format transaction).
+
+[TODO: if the non-upgraded chain only had empty blocks, technically that would be valid under new rules, so
+could cause a wipe-out. Fixing this would require at least changing the block version, in which case blocks
+may as well commit to a ``BRANCH_ID``.]
```

Comment:
Empty blocks have coinbase transactions, even without input signatures,  should violate the new transaction format used by Overwinter, so when verifying the empty blocks they should be rejected when the tx is deemed invalid.

Reply:
This is true for Overwinter, but may not be true in general. I will change the TODO to a note about this.